rust/src/collaboration/databasesync.rs

```diff
@@ -0,0 +1,708 @@
+/// * `name_changeset` - Function to call for naming a pushed changeset, if necessary
+pub fn upload_database<P: ProgressCallback, N: NameChangeset>(
+    metadata: &FileMetadata,
+    project: &RemoteProject,
+    parent_folder: Option<&RemoteFolder>,
+    mut progress: P,
+    mut name_changeset: N,
+) -> Result<RemoteFile, ()> {
+    let folder_raw = parent_folder.map_or(ptr::null_mut(), |h| unsafe { h.as_raw() } as *mut _);
+    let result = unsafe {
+        BNCollaborationUploadDatabase(
+            metadata.handle,
+            project.as_raw(),
+            folder_raw,
+            Some(P::cb_progress_callback),
+            &mut progress as *mut P as *mut ffi::c_void,
+            Some(N::cb_name_changeset),
+            &mut name_changeset as *mut N as *mut ffi::c_void,
+        )
+    };
+    ptr::NonNull::new(result)
+        .map(|raw| unsafe { RemoteFile::from_raw(raw) })
+        .ok_or(())
+}
+
+/// Test if a database is valid for use in collaboration
+pub fn is_collaboration_database(database: &Database) -> bool {
+    unsafe { BNCollaborationIsCollaborationDatabase(database.as_raw()) }
+}
+
+/// Get the Remote for a Database
+pub fn get_remote_for_local_database(database: &Database) -> Result<Option<Remote>, ()> {
+    let mut value = ptr::null_mut();
+    let success =
+        unsafe { BNCollaborationGetRemoteForLocalDatabase(database.as_raw(), &mut value) };
+    success
+        .then(|| ptr::NonNull::new(value).map(|handle| unsafe { Remote::from_raw(handle) }))
+        .ok_or(())
+}
+
+/// Get the Remote for a BinaryView
+pub fn get_remote_for_binary_view(bv: &BinaryView) -> Result<Option<Remote>, ()> {
+    let Some(db) = bv.file().database() else {
+        return Ok(None);
+    };
+    get_remote_for_local_database(&db)
+}
+
+/// Get the Remote Project for a Database, returning the Remote project from one of the
+/// connected remotes, or None if not found or if projects are not pulled
+pub fn get_remote_project_for_local_database(
+    database: &Database,
+) -> Result<Option<RemoteProject>, ()> {
+    let mut value = ptr::null_mut();
+    let success =
+        unsafe { BNCollaborationGetRemoteProjectForLocalDatabase(database.as_raw(), &mut value) };
+    success
+        .then(|| ptr::NonNull::new(value).map(|handle| unsafe { RemoteProject::from_raw(handle) }))
+        .ok_or(())
+}
+
+/// Get the Remote File for a Database
+pub fn get_remote_file_for_local_database(database: &Database) -> Result<Option<RemoteFile>, ()> {
+    let mut value = ptr::null_mut();
+    let success =
+        unsafe { BNCollaborationGetRemoteFileForLocalDatabase(database.as_raw(), &mut value) };
+    success
+        .then(|| ptr::NonNull::new(value).map(|handle| unsafe { RemoteFile::from_raw(handle) }))
+        .ok_or(())
+}
+
+/// Add a snapshot to the id map in a database
+pub fn assign_snapshot_map(
+    local_snapshot: &Snapshot,
+    remote_snapshot: &CollabSnapshot,
+) -> Result<(), ()> {
```

Comment:
`Result<(), ()>` -> `bool`

Reply:
The `Result` can't be ignored by default, at least not without a warning, but the `bool` can, so we want to force the user to consider the returned value.